lib/speed.js

```diff
@@ -1,35 +1,105 @@
 'use strict';
 
-const rpad = (s, char, count) => s + char.repeat(count - s.length);
-const lpad = (s, char, count) => char.repeat(count - s.length) + s;
+const NS_PER_SEC = 1e9;
 
-const speed = (caption, count, tests) => {
-  const times = tests.map(fn => {
-    const result = [];
-    const begin = process.hrtime();
-    for (let i = 0; i < count; i++) result.push(fn());
-    const end = process.hrtime(begin);
-    const time = end[0] * 1e9 + end[1];
-    const name = rpad(fn.name, '.', 25);
-    return { name, time };
+const roundTo = (num, decimals) => {
+  const c = 10 ** decimals;
+  return Math.round(num * c) / c;
+};
+
+const run = (count, fn, args) => {
+  let result = null;
+  const start = process.hrtime();
+  for (let i = 0; i < count; i++) {
+    // eslint-disable-next-line prefer-spread
+    const r = fn.apply(null, args);
+    if (Math.random() > 0.5) result = r;
+  }
+  const elapsed = process.hrtime(start);
+  const time = elapsed[0] * NS_PER_SEC + elapsed[1];
+  return { result, args, count, time };
+};
+
+// Microbenchmark each passed configuration multiple times and compare results.
+//   cases <Array> cases to test, each case contains
+//     fn <Function> function to check, will be called with each args provided
+//     name <string> case name, function.name by default
+//     args <Array>
+//     n <number>
+const measure = (
+  cases,
+  { defaultCount = 1e6, preflight = 10, runs = 20, listener = {} } = {}
+) => {
+  const res = [];
+  cases.forEach(({ name, fn, argCases = [], n = defaultCount }) => {
+    if (!n || n <= 0) throw new Error('Run count not specified');
+    const testName = name || fn.name || 'anonymous';
+    if (argCases.length === 0) argCases.push(undefined);
+    if (listener.preflight) listener.preflight(testName, argCases[0]);
+    for (let i = 0; i < preflight; i++) run(n, fn, argCases[0]);
```

Comment:
Wouldn't it be better to use every argument?

Reply:
I guess but I don't think it will provide much benefit while making running time significantly longer. We should probably add separate runcount for preflight, though I'd prefer to do that in a separate PR.

Reply:
Done.